File: app/src/main/java/com/fastaccess/provider/timeline/handler/drawable/GlideDrawableTarget.java
package com.fastaccess.provider.timeline.handler.drawable;

import android.graphics.Rect;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.widget.TextView;

import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.fastaccess.R;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

class GlideDrawableTarget extends SimpleTarget<Drawable> {
    private final UrlDrawable urlDrawable;
    private final WeakReference<TextView> container;
    private final int width;

    GlideDrawableTarget(UrlDrawable urlDrawable, WeakReference<TextView> container, int width) {
        this.urlDrawable = urlDrawable;
        this.container = container;
        this.width = width;
    }

    @Override public void onResourceReady(final @NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
        if (container != null && container.get() != null) {
            TextView textView = container.get();
            textView.post(() -> {
                float width;
                float height;
                if (resource.getIntrinsicWidth() >= this.width) {
                    float downScale = (float) resource.getIntrinsicWidth() / this.width;
                    width = (float) (resource.getIntrinsicWidth() / downScale / 1.3);
                    height = (float) (resource.getIntrinsicHeight() / downScale / 1.3);
                } else {
                    width =  (float) resource.getIntrinsicWidth();
                    height = (float) resource.getIntrinsicHeight();
                }
                Rect rect = new Rect(0, 0, Math.round(width), Math.round(height));
                resource.setBounds(rect);
                urlDrawable.setBounds(rect);
                urlDrawable.setDrawable(resource);
                if (resource instanceof GifDrawable) {
                    urlDrawable.setCallback((Drawable.Callback) textView.getTag(R.id.drawable_callback));
                    ((GifDrawable) resource).setLoopCount(GifDrawable.LOOP_FOREVER);
                    ((GifDrawable) resource).start();
                }
                textView.setText(textView.getText());
                textView.invalidate();
            });
        }
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/spans/drawable/GlideDrawableTarget.kt
package com.fastaccess.markdown.spans.drawable

import android.annotation.SuppressLint
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.Animatable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import androidx.appcompat.graphics.drawable.DrawableWrapper
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import timber.log.Timber

@SuppressLint("RestrictedApi")
class GlideDrawableTarget(private val width: Int) : SimpleTarget<Drawable>() {


    private val transparentDrawable = ColorDrawable(Color.TRANSPARENT)
    private val wrapper = DrawableWrapper(null)

    val lazyDrawable get() = wrapper

    init {
        setDrawable(null)
    }

    override fun onLoadStarted(placeholder: Drawable?) {
        setDrawable(placeholder)
    }


    override fun onLoadFailed(errorDrawable: Drawable?) {
        setDrawable(errorDrawable)
    }


    override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
        Timber.e("${resource is Animatable}---${resource is GifDrawable}")
        if (resource is GifDrawable) {
            resource.setLoopCount(GifDrawable.LOOP_FOREVER)
            resource.start()
        }
        setDrawable(resource)
    }


    override fun onLoadCleared(placeholder: Drawable?) {
        setDrawable(placeholder)
    }

    private fun setDrawable(resource: Drawable?) {
        var drawable = resource
        if (drawable == null) {
            drawable = transparentDrawable
        }
        val width: Float
        val height: Float
        if (drawable.intrinsicWidth >= this.width) {
            val downScale = drawable.intrinsicWidth.toFloat() / this.width
            width = (drawable.intrinsicWidth.toDouble() / downScale.toDouble() / 1.3).toFloat()
            height = (drawable.intrinsicHeight.toDouble() / downScale.toDouble() / 1.3).toFloat()
        } else {
            val multiplier = this.width.toFloat() / drawable.intrinsicWidth
            width = drawable.intrinsicWidth.toFloat() * multiplier
            height = drawable.intrinsicHeight.toFloat() * multiplier
        }
        val rect = Rect(0, 0, width.toInt(), height.toInt())
        wrapper.wrappedDrawable = drawable
        wrapper.bounds = rect
        drawable.bounds = wrapper.bounds
        wrapper.invalidateSelf()
    }
}